### PR TITLE
Two Factor Authentication endpoints

### PR DESCRIPTION
**Pull request by CityVibes**
Implemented calls to 2fa endpoints + wrapper method to authenticate by passing in mfa & otp.

### CODE REVIEW

**Review by RaimundasSakalauskas — APPROVED** [review]
Seems great. I might add "getMFAToken" to iOS too as this seems like a nice helper method.

File: cloudsdk/src/main/java/io/particle/android/sdk/cloud/exceptions/ParticleCloudException.java [hunk 1]
```diff
@@ -154,6 +155,26 @@ public String getServerErrorMsg() {
         return serverErrorMessage;
     }
 
+    /**
+     * Server-provided multi-factor auth token.  May be null.
+     *
+     * @return server-provided multi-factor auth token or null
+     */
+    public String getMfaToken() {
```
Comment (jensck) [inline cloudsdk/src/main/java/io/particle/android/sdk/cloud/exceptions/ParticleCloudException.java]: We're already using GSON everywhere else. Why the inconsistency?

Comment (jensck) [reply cloudsdk/src/main/java/io/particle/android/sdk/cloud/exceptions/ParticleCloudException.java]: Adding a specialized auth method onto a generic API exception doesn't feel great to me.  Could this specialized logic be handled elsewhere (somewhere that isn't exposed as public API surface)?

**Review by jensck — CHANGES_REQUESTED** [review]
If iOS has a `getMFAToken()`, either we should have it, too, or we should eliminate it from both.

I lean toward the latter.  Do we really want this as public API?  How would it be used for _anything_ but the login code that we're already providing?

**Comment by CityVibes** [comment]
Just copying what I said on slack regarding mfa getter "I would say - yes, if we have public API for login why wouldn't we have public mfa token getter it's part of login flow after all. Shouldn't user be able to implement same behavior as our own tinker app?"

File: cloudsdk/src/main/java/io/particle/android/sdk/cloud/exceptions/ParticleCloudException.java (hunk 1 above)
Comment (CityVibes) [reply cloudsdk/src/main/java/io/particle/android/sdk/cloud/exceptions/ParticleCloudException.java]: Note 'loadServerErrorMsg' method, we infact do not use gson here, so this is consistent with the rest of the class.

Comment (CityVibes) [reply cloudsdk/src/main/java/io/particle/android/sdk/cloud/exceptions/ParticleCloudException.java]: I believe it has value as public API, we could however handle it elsewhere or have child exception, something like 'ParticleLoginException'.

**Comment by CityVibes** [comment]
Moved MFA getter to new ParticleLoginException (inherits from ParticleCloudException), this way we won't expose getter with exceptions unrelated to login.

**Comment by jensck** [comment]
+1 for the specialized `ParticleLoginException`

**Pull request merged**